src/vsr/grid.zig

```diff
@@ -264,11 +276,14 @@ pub fn GridType(comptime Storage: type) type {
         fn checkpoint_join(grid: *Grid) void {
             assert(grid.checkpointing != null);
             assert(grid.canceling == null);
+            assert(grid.read_global_queue.empty());
             grid.assert_only_repairing();
 
-            // TODO(Grid repair queue): This check is too conservative – we only need to wait for
-            // all released-block writes to finish.
-            if (grid.write_iops.executing() > 0) return;
+            // We are still repairing some blocks that were released at the checkpoint.
+            if (!grid.repair_queue.checkpoint_complete()) {
+                assert(grid.write_iops.executing() > 0);
+                return;
+            }
```

Comment:
Before: Any in-progress repairs (i.e. syncing blocks) would stall the checkpoint until they finish.
Now: We are more precise -- the checkpoint only needs to stall for writes to blocks that are about to be released.

(The goal is that we never write to a block _after_ it has been released, since we may re-acquire and re-write it in the new checkpoint.)